### PR TITLE
Initial parser implementation

### PR DESCRIPTION
**Pull request by yazgazan**
TODO:

- [x] write parser tests
- [x] better error handling

### CODE REVIEW

File: ast/ast.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	node_type_literal_string
+} e_node_type;
+
+typedef enum {
+	value_type_string,
+	value_type_integer,
+	value_type_undefined
+} e_value_type;
+
+typedef union {
```
Comment (notoriou1pig) [inline ast/ast.h]: Smart, might be (void *data) then cast it.

Comment (yazgazan) [reply ast/ast.h]: I removed it to simplify things (it wasn't being used). We can put it back when we need to handle types other than strings.

File: ast/ast.c [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+#include "ast.h"
+
+static char *copystr(char *src);
+
+t_node *new_node(e_node_type type, char *data)
+{
+	t_node *node;
+
+	node = calloc(1, sizeof(*node));
```
Comment (notoriou1pig) [inline ast/ast.c:14]: Why calloc ? bzero everything ?

Comment (yazgazan) [reply ast/ast.c:14]: Yep, this avoids some initialisation.

File: states/states.c [hunk 3]
```diff
@@ -8,7 +8,7 @@
 extern char **environ;
```
Comment (notoriou1pig) [inline states/states.c:8]: What is this variable ? It's not the first time I see it.
Is it the environment ?

Comment (yazgazan) [reply states/states.c:8]: Indeed it is: `man 7 environ` (http://man7.org/linux/man-pages/man7/environ.7.html)

**Pull request merged**